### PR TITLE
JSON Diff

### PR DESCRIPTION
**Pull request by nablaone**
This PR contains a JSON comparator. This comparator returns results as a list of problems (highly inspired by https://datatracker.ietf.org/doc/html/rfc6902).  

It's not finished. There are some TODOs in the code. It will be solved during the real first A/B testing session.

### CODE REVIEW

File: quesma/jsondiff/jsondiff.go [hunk 1]
```diff
@@ -0,0 +1,253 @@
...
+}
+
+var (
+	invalidType        = newType("invalid_type", "Types are not equal")
+	invalidValue       = newType("invalid_value", "Values are not equal")
+	invalidArrayLength = newType("invalid_array_length", "Array lengths are not equal")
+	objectDifference   = newType("object_difference", "Objects are different")
+)
+
+type Problem struct {
```
Comment (jakozaur) [inline quesma/jsondiff/jsondiff.go:30]: Too generic. `ResponseDiff`?

Comment (jakozaur) [reply quesma/jsondiff/jsondiff.go:30]: `JsonMismatch`, `ResponseMismatch`

File: quesma/jsondiff/jsondiff_test.go [hunk 2]
```diff
@@ -0,0 +1,127 @@
...
+		{
+			name:     "deep path difference",
+			expected: `{"a": {"d": {"b": 1}}, "c": 3}`,
+			actual:   `{"a": {"d": {"b": 2}}, "c": 3}`,
+			problems: []Problem{problem("a.d.b", invalidValue)},
+		},
+
+		{
+			name:     "deep path difference",
+			expected: `{"a": {"d": {"b": 1}}, "c": 3, "_ignore": 1}`,
```
Comment (pdelewski) [inline quesma/jsondiff/jsondiff_test.go:91]: What with the case where one json has more keys than the other? Should it be reported?

Comment (nablaone) [reply quesma/jsondiff/jsondiff_test.go:91]: Not all corner cases are covered in this PR.  I will fix it in the next ones.

File: quesma/jsondiff/jsondiff.go (hunk 1 above)
Comment (nablaone) [reply quesma/jsondiff/jsondiff.go]: Renamed to JSONMismatch.

File: quesma/jsondiff/jsondiff.go [hunk 3]
```diff
@@ -0,0 +1,253 @@
...
+func (p JSONMismatch) String() string {
+	return fmt.Sprintf("%s: %s, expected: %s, actual: %s", p.Path, p.Message, p.Expected, p.Actual)
+}
+
+type Mismatches []JSONMismatch
+
+func (r Mismatches) String() string {
+	var s string
+	for _, p := range r {
+		s += p.String() + "\n"
```
Comment (jakozaur) [inline quesma/jsondiff/jsondiff.go:51]: Maybe next PR, but `strings.Builder` is way more efficient.

File: quesma/jsondiff/jsondiff.go [hunk 4]
```diff
@@ -0,0 +1,253 @@
...
+		keys = append(keys, k)
+	}
+
+	sort.Strings(keys)
+
+	return keys
+}
+
+func (d *JSONDiff) intersect(a, b []string) []string {
+	var c []string
```
Comment (jakozaur) [inline quesma/jsondiff/jsondiff.go:128]: Add a comment if it's not efficient; otherwise, sort it and then compare.

**Review by jakozaur — APPROVED** [review]

**Pull request merged**